url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/manager/FPMMusicAgent.java

```diff
@@ -0,0 +1,332 @@
+	}
+
+	public void msgStartMessageTone() {
+		startMessageTone = true;
+		stateChanged();
+	}
+
+	/*
+	 * Scheduler (non-Javadoc)
+	 * 
+	 * @see agent.Agent#pickAndExecuteAnAction()
+	 */
+	@Override
+	public boolean pickAndExecuteAnAction() {
+		print("In my scheduler");
+		if (!sfxActive()) {
+			stopSFX();
+			music.start();
+		}
+		if (startFlute && !sfxActive()) {
+			startFlute = false;
+			startPokeflute();
+			return true;
+		}
+
+		if (startCompleted && !sfxActive()) {
+			startCompleted = false;
+			startCompleted();
+			return true;
+		}
+
+		if (startRecovery && !sfxActive()) {
+			startRecovery = false;
+			startRecovery();
+			return true;
+		}
+
+		if (startMessageTone) {
+			startMessageTone = false;
+			startMessageTone();
+			return true;
+		}
+		return false;
+	}
+
+	/*
+	 * Actions
+	 */
+	public void startMusic() {
+		print("Playing music");
+		// stopCompleted();
+		// stopPokeflute();
+		// stopRecovery();
+
+		if (music != null) {
+			//if (!music.isActive() && !sfxActive()) {
+				music.loop(Clip.LOOP_CONTINUOUSLY);
+				//music.start);
+			//}
+		}
+		stateChanged();
+	}
+
+	public void stopSFX() {
+		pokeflute.stop();
+		pokeflute.flush();
+		completed.stop();
+		completed.flush();
+		recovery.stop();
+		recovery.flush();
+	}
+
+	public boolean sfxActive() {
+		return pokeflute.isActive() || completed.isActive() || recovery.isActive();
+	}
+
+	public void startPokeflute() {
+		print("Playing pokeflute");
+		if (pokeflute != null) {
+			try {
+				musicSem.acquire();
+			} catch (InterruptedException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
+			}
+			music.stop();
+			music.flush();
+			// stopCompleted();
+			// stopRecovery();
+
+			System.out.println("plays flute"); // !!! EXTREMELY IMPORTANT
+
+			stopSFX();
+			// pokeflute.stop();
+			pokeflute.setFramePosition(0);
+			pokeflute.start();
+			musicTimer.schedule(new TimerTask() {
+				@Override
+				public void run() {
+					startMusic();
+				}
+			}, 4224);
+
+			// while (pokeflute.isRunning()) {
+			// ;
+			// }
+			musicSem.release();
+			//startMusic();
+		}
+		stateChanged();
+	}
+
+	public void startRecovery() {
+		print("Playing recovery");
+		if (recovery != null) {
+			try {
+				musicSem.acquire();
+			} catch (InterruptedException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
+			}
+
+			music.stop();
+			music.flush();
+			// stopPokeflute();
+			// stopCompleted();
+			System.out.println("plays recovery"); // !!! EXTREMELY IMPORTANT
+
+			stopSFX();
+			// recovery.stop();
+			recovery.setFramePosition(0);
+			recovery.start();
+			musicTimer.schedule(new TimerTask() {
+				@Override
+				public void run() {
+					startMusic();
+				}
+			}, 2000);
+
+			musicSem.release();
+			//startMusic();
+		}
+		stateChanged();
+	}
+
+	public void startCompleted() {
+		print("Playing completed");
+		if (completed != null) {
+			try {
+				musicSem.acquire();
+			} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)